File: java/src/br/com/brunohpaiva/osprocessscheduling/algorithms/Sjf.java
package br.com.brunohpaiva.osprocessscheduling.algorithms;

import br.com.brunohpaiva.osprocessscheduling.comparators.ArrivalTimeComparator;
import br.com.brunohpaiva.osprocessscheduling.comparators.BurstTimeComparator;
import br.com.brunohpaiva.osprocessscheduling.models.Process;

import java.util.Comparator;
import java.util.List;

public class Sjf {

    public static Double calculate(List<Process> processes, Comparator<Process> tiebreakerComparator) {
        processes.sort(new ArrivalTimeComparator(tiebreakerComparator));

        Comparator<Process> burstTimeComparator = new BurstTimeComparator(tiebreakerComparator);
        int currentExecutionTime = 0;
        double totalWaitingTime = 0.0;
        Process process = processes.get(0);

        while (process != null) {
            int executedAt = currentExecutionTime;
            currentExecutionTime += process.getBurstTime();
            int waitingTime = executedAt - process.getArrivalTime();
            totalWaitingTime += waitingTime;

            process.setFinished(true);
            process = getNextProcess(processes, currentExecutionTime, burstTimeComparator);
        }

        return totalWaitingTime / processes.size();
    }

    public static Process getNextProcess(List<Process> processes, int currentExecutionTime, Comparator<Process> burstTimeComparator) {
        return processes.stream().filter(process -> process.getArrivalTime() <= currentExecutionTime && !process.isFinished()).min(burstTimeComparator).orElse(null);
    }

}

File: typescript/src/algorithms/sjf.ts
import createArrivalTimeComparator from "../comparators/arrivalTime";
import createBurstTimeComparator from "../comparators/burstTime";
import { AlgorithmFunction, ProcessComparator } from "../types";
import Process from "../models/Process";

const getNextProcess = (
  processes: Process[],
  currentExecutionTime: number,
  tiebreakerComparator?: ProcessComparator
): Process => {
  let burstTimeComparator = createBurstTimeComparator(tiebreakerComparator);
  return processes
    .filter(
      (process): boolean =>
        process.arrivalTime <= currentExecutionTime && !process.finished
    )
    .sort(burstTimeComparator)
    .shift() as Process;
};

const sjf: AlgorithmFunction = (processes, options = {}): number => {
  processes = processes
    .map((process): Process => process.clone())
    .sort(createArrivalTimeComparator(options.tiebreakerComparator));

  let currentExecutionTime = 0;
  let totalWaitingTime = 0;
  let process = processes[0];

  while (process) {
    let executedAt = currentExecutionTime;
    currentExecutionTime += process.burstTime;
    let waitingTime = executedAt - process.arrivalTime;
    totalWaitingTime += waitingTime;

    process.finished = true;
    process = getNextProcess(
      processes,
      currentExecutionTime,
      options.tiebreakerComparator
    );
  }

  return totalWaitingTime / processes.length;
};

export default sjf;
